### PR TITLE
Logging: All-up logging improvements:

### LINKED ISSUE #1374: Flag to allow application data in logging
From a security point of view we decided that, by default, both the message action and state passed to ILogger.Write should not contain any application data. That is, data that comes from the data store or may be supplied by an end user of the application. Depending on the application, this data may contain sensitive information (usernames, credit card numbers, etc.). Examples include, results from queries, values stored in entity instances, and constant values used in LINQ expressions.

We'll have a simple flag that can be enabled to allow this information to be included. This is especially useful for logging queries, where it is helpful to have the parameter values logged along with the query.

This work item includes:
- [x] Creating the flag (off by default)
- [ ] Reviewing existing logging to make sure it honors the flag
- [ ] Writing a warning to the log when the flag is enabled (once per context, ServiceProvider, or whatever is easiest)
- [ ] Wherever data is filtered out of a message, include a message in place of the data letting folks know they can enable the flag to get the data (and a warning that it may result in logging sensitive info) 

We should try and add this in a way that makes it easy to use it for future logging that we add.

### PR DESCRIPTION
**Pull request by anpete**
- New DbContextOption "LogSqlParameterValues()" replacing use of Debug log level to control sensitive data logging.
- Use a wrapping ILogger<T> in relational to transparently enforce sensitive data logging.
- More consistentizing of logging usage, DI etc.
- Better structured logging of DbCommands.
- Use assembly scoped enums for event ids.
- Use event ids everywhere.

Fix #1374, #2490

### CODE REVIEW

File: src/EntityFramework.Relational/Infrastructure/RelationalEntityFrameworkServicesBuilderExtensions.cs [hunk 1]
```diff
@@ -110,4 +114,54 @@ private static IRelationalDatabaseProviderServices GetProviderServices(IServiceP
             return providerServices;
         }
     }
+
+    public interface ILogSqlParameterValuesAccessor
```
Comment (ajcvickers) [inline src/EntityFramework.Relational/Infrastructure/RelationalEntityFrameworkServicesBuilderExtensions.cs]: Move to its own file?

File: src/EntityFramework.Relational/Infrastructure/RelationalEntityFrameworkServicesBuilderExtensions.cs [hunk 2]
```diff
@@ -110,4 +114,54 @@ private static IRelationalDatabaseProviderServices GetProviderServices(IServiceP
...
+    {
+        bool LogSqlParameterValues { get; }
+}
+
+    public class SensitiveDataLogger<T> : ILogger<T>, ILogSqlParameterValuesAccessor
```
Comment (ajcvickers) [inline src/EntityFramework.Relational/Infrastructure/RelationalEntityFrameworkServicesBuilderExtensions.cs]: Move to its own file?

File: src/EntityFramework.Relational/Infrastructure/RelationalEntityFrameworkServicesBuilderExtensions.cs [hunk 3]
```diff
@@ -32,7 +34,9 @@ public static EntityFrameworkServicesBuilder AddRelational([NotNull] this Entity
         {
             Check.NotNull(builder, nameof(builder));
 
-            builder.GetService().TryAdd(new ServiceCollection()
+            builder.GetService()
+                .Replace(new ServiceDescriptor(typeof(ILogger<>), typeof(SensitiveDataLogger<>), ServiceLifetime.Scoped))
```
Comment (ajcvickers) [inline src/EntityFramework.Relational/Infrastructure/RelationalEntityFrameworkServicesBuilderExtensions.cs]: Does this working depend on when AddEntityFramework is called? In other words, can I accidentally defeat this if I do some logging registration in Startup after AddEntityFramework has already been called?

If so, do we care?

**Comment by ajcvickers** [comment]
:shipit:

File: src/EntityFramework.Relational/Infrastructure/RelationalEntityFrameworkServicesBuilderExtensions.cs (hunk 1 above)
Comment (anpete) [inline src/EntityFramework.Relational/Infrastructure/RelationalEntityFrameworkServicesBuilderExtensions.cs]: Will do.

File: src/EntityFramework.Relational/Infrastructure/RelationalEntityFrameworkServicesBuilderExtensions.cs (hunk 2 above)
Comment (anpete) [inline src/EntityFramework.Relational/Infrastructure/RelationalEntityFrameworkServicesBuilderExtensions.cs]: Will do.

File: src/EntityFramework.Relational/Infrastructure/RelationalEntityFrameworkServicesBuilderExtensions.cs (hunk 3 above)
Comment (anpete) [inline src/EntityFramework.Relational/Infrastructure/RelationalEntityFrameworkServicesBuilderExtensions.cs]: Discussed offline and decided to introduce a new type of ILogger instead.

**Pull request closed without merging**